File: hard/maximalrectangle.cpp
#include <iostream>

class Solution {
public:
    int maximalRectangle(vector<vector<char>>& matrix) {   
        stack<int> stack;
        vector<int> temp(matrix[0].size(), 0);
        int area = 0;

        for(int i = 0; i < matrix.size(); i++) {
            for(int j = 0; j < matrix[i].size(); j++) {
                if(matrix[i][j] == '1') {
                    temp[j] += 1; 
                } else {
                    temp[j] = 0;
                }
            }

            int m = temp.size();
            for(int k = 0; k < m; k++) {
                while(!stack.empty() && temp[stack.top()] > temp[k]) {
                    int h = temp[stack.top()];
                    stack.pop();
                    int w = stack.empty() ? k : k - stack.top() - 1;
                    area = max(area, w * h);
                }
                stack.push(k);
            }

            while(!stack.empty()) {
                int h = temp[stack.top()];
                stack.pop();
                int w = stack.empty() ? m : m - stack.top() - 1;
                area = max(area, w * h);
            }
        }

        return area;
    }
};

File: python/hard/maximalrectangle.py
class Solution:
    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        if not matrix:
            return 0

        rows, cols = len(matrix), len(matrix[0])
        temp = [0] * (cols + 1)
        max_area = 0

        for i in range(rows):
            for j in range(cols):
                if matrix[i][j] == '1':
                    temp[j] += 1
                else:
                    temp[j] = 0

            stk = []
            for k in range(len(temp)):
                while stk and temp[stk[-1]] > temp[k]:
                    h = temp[stk.pop()]
                    w = k if not stk else k - stk[-1] - 1
                    max_area = max(max_area, h * w)
                stk.append(k)

        return max_area
